File: email-aggregator-go/src/model/canonical.go
// Package model 定义邮箱聚合平台的领域模型与统一契约（ADR-005）。
// 本包不依赖任何其它业务包，是 connector / sync / events / security 的公共基座。
package model

// Provider 邮件服务商协议类型
type Provider string

const (
	ProviderIMAP       Provider = "imap"
	ProviderPOP3       Provider = "pop3"
	ProviderExchange   Provider = "exchange"
	ProviderGmail      Provider = "gmail"
	ProviderEnterprise Provider = "enterprise"
)

// Address 邮件地址（RFC 5322）
type Address struct {
	Name  string `json:"name"`
	Email string `json:"email"`
}

// AttachmentMeta 附件元信息（内容本身存对象存储，这里仅索引）
type AttachmentMeta struct {
	Filename    string `json:"filename"`
	ContentType string `json:"contentType"`
	SizeBytes   int64  `json:"sizeBytes"`
	ContentHash string `json:"contentHash"`
}

// SyncCursor 断点续传游标（协议相关字段可选，多协议归一化）
type SyncCursor struct {
	UIDValidity      uint32            `json:"uidValidity,omitempty"`
	UIDNext          uint32            `json:"uidNext,omitempty"`
	LastUID          uint32            `json:"lastUid,omitempty"`
	ModSeq           uint64            `json:"modseq,omitempty"`
	HighWaterMark    string            `json:"highWaterMark,omitempty"`
	ProviderSpecific map[string]string `json:"providerSpecific,omitempty"`
}

// CanonicalMail 统一邮件模型（ADR-005 归一化产物，所有协议收敛于此）
type CanonicalMail struct {
	TenantID      string           `json:"tenantId,omitempty"` // ADR-009 多租户：数据面租户归属（存储/检索/通知全链路透传）
	ID            string           `json:"id"`
	AccountID     string           `json:"accountId"`
	Provider      Provider         `json:"provider"`
	Folder        string           `json:"folder"`
	From          Address          `json:"from"`
	To            []Address        `json:"to"`
	Cc            []Address        `json:"cc"`
	Bcc           []Address        `json:"bcc"`
	Subject       string           `json:"subject"`
	BodyText      string           `json:"bodyText,omitempty"`
	BodyHTML      string           `json:"bodyHTML,omitempty"`
	Snippet       string           `json:"snippet,omitempty"`
	HasAttachment bool             `json:"hasAttachment"`
	Attachments   []AttachmentMeta `json:"attachments,omitempty"`
	InternalDate  int64            `json:"internalDate"`
	SizeBytes     int64            `json:"sizeBytes"`
	Cursor        SyncCursor       `json:"cursor"`
	Read          bool             `json:"read"`                   // 用户已读状态（CRUD 生命周期）
	RawObjectKey  string           `json:"rawObjectKey,omitempty"` // 内容寻址对象存储引用
}

File: email-aggregator/src/model/canonical.ts
/**
 * 统一邮件领域模型（Canonical Mail Model）。
 * 这是平台内部唯一表示，屏蔽 IMAP / POP3 / Exchange(Graph) / Gmail 协议差异。
 * 适配器负责：协议 ↔ CanonicalMail 的双向映射。
 */

export type FolderName =
  | "INBOX"
  | "SENT"
  | "DRAFTS"
  | "TRASH"
  | "JUNK"
  | "CUSTOM";

export interface Contact {
  name?: string;
  address: string; // RFC5322 addr-spec
}

export interface Attachment {
  filename: string;
  mimeType: string;
  sizeBytes: number;
  /** 内容寻址哈希（sha256），用于去重：相同附件只存一份 */
  contentHash: string;
  /** 内容存储中的对象键（由 content store 回填） */
  objectKey?: string;
}

export interface MailFlags {
  seen: boolean;
  flagged: boolean;
  answered: boolean;
  deleted: boolean;
}

/**
 * 同步游标：各协议用自己的字段，统一封装。
 * 仅适配层关心，不进入检索模型。
 */
export interface SyncCursor {
  uidValidity?: string; // IMAP
  uidNext?: string; // IMAP
  highestModSeq?: string; // IMAP CONDSTORE / QRESYNC
  historyId?: string; // Gmail
  deltaLink?: string; // Graph (Microsoft)
  uidlWatermark?: string; // POP3
}

export interface CanonicalMail {
  /** 幂等去重键 = (tenantId, accountId, folder, messageId) */
  idempotencyKey: string;
  /** 租户 ID（ADR-009）：逻辑多租户隔离主键，空值由下游 Resolve 收敛到 DefaultTenantID */
  tenantId?: string;
  accountId: string;
  folder: FolderName;
  /** RFC822 Message-ID */
  messageId: string;
  from: Contact[];
  to: Contact[];
  cc: Contact[];
  subject: string;
  bodyText?: string;
  bodyHtml?: string;
  attachments: Attachment[];
  flags: MailFlags;
  /** 收件时间戳（ms） */
  internalDate: number;
  sizeBytes: number;
  /** 协议侧游标（断点续传） */
  cursor: SyncCursor;
  /** 内容存储中的原始 MIME 对象键（用于从对象存储还原正文/附件） */
  rawObjectKey?: string;
}

export interface Folder {
  name: FolderName;
  /** 统一文件夹路径（多层级以 '/' 分隔） */
  path: string;
  messageCount: number;
  unreadCount: number;
}

/** 上游返回的"变化集"：新增 / 变更 / 删除的 UID 列表 */
export interface ChangeSet {
  added: string[];
  changed: string[];
  deleted: string[];
  /** 本批次后的最新游标，供持久化 */
  nextCursor: SyncCursor;
}

/**
 * 幂等去重键构造。
 * - 单租户（无 tenantId）：`${accountId}|${folder}|${messageId}` —— 与改造前完全等价（最小可逆）。
 * - 多租户（带 tenantId）：`${tenantId}|${accountId}|${folder}|${messageId}` —— 防止跨租户邮件碰撞。
 * 与 Go 实现对齐：email-aggregator-go 幂等键 `tid:accountId:mailId` 同义（分隔符为 TS 历史约定）。
 */
export function idempotencyKey(
  accountId: string,
  folder: FolderName,
  messageId: string,
  tenantId?: string,
): string {
  return tenantId
    ? `${tenantId}|${accountId}|${folder}|${messageId}`
    : `${accountId}|${folder}|${messageId}`;
}
